tools/shared/src/main/scala/org/scalajs/core/tools/javascript/IncClassEmitter.scala

```diff
@@ -0,0 +1,61 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2016, LAMP/EPFL        **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package org.scalajs.core.tools.javascript
+
+import scala.collection.mutable
+
+import org.scalajs.core.tools.sem._
+
+import org.scalajs.core.tools.linker.LinkingUnit
+
+/** Tracks incrementality breaking changes made during class emitting.
+ */
+final class IncClassEmitter(semantics: Semantics, outputMode: OutputMode) {
+
+  private var currentClassesCtorOpt: Set[String] = Set.empty
+  private[this] val askedForCtorOpt =
+    mutable.Map.empty[String, Set[(String, String, Boolean)]]
+
+  def beginRun(unit: LinkingUnit,
+      invalidateFunc: (String, String, Boolean) => Unit): ScalaJSClassEmitter = {
+    val classEmitter = new ScalaJSClassEmitter(outputMode, unit, this)
+    val newClassesCtorOpt = if (outputMode == OutputMode.ECMAScript6StrongMode) {
+      Set.empty[String]
+    } else {
+      classEmitter.candidateForJSConstructorOpt
+    }
```

Comment:
I think this design needs some rethinking if we always need two instances of `ScalaJSClassEmitter` if we want to emit something...
I would consider either of the following:

1. Track trivial properties (isInstantiated, superClass, etc.)
2. Inject the properties you want to query for in a `beginRun`

I'd also consider having "alive" objects in the emitter (essentially the caches), that register themselves. Tracking this by names is going to be terrible. (And what if two things of different types depend on the same knowledge?).